### ISSUE TITLE
[question] NDK request permissions

### ISSUE THREAD

**Issue by Gibson85** [android-ndk/ndk#1031]
Hi, I couldn't find a solution for a pure NDK app to request permissions e.g. to read/write on external storage. What is the intended way to do so from C++?

**Comment by DanAlbert** [comment]
You can't. You have to make the call through JNI. If you're just trying to avoid writing Java, you can look up the Java classes and methods from C++ through JNI like anything else and make the calls that way, but there's no such thing as a "pure NDK app". You're still running within a Java process and most APIs are Java only.

**Comment by Gibson85** [comment]
Yes, I expected this, but I did not got it working so far.
The code below result in return code "-1" uknown error.
What am I doing wrong?

```
static int InitJVM(JavaVM** p_vm, JNIEnv** p_env)
{
	LOGI("JNI_CreateJavaVM...");

	JavaVMOption options[1];
	memset(&options, 0, sizeof(options));

	options[0].optionString = "-Djava.class.path=.";
	//options[1].optionString = "-Xms1m"; // 1MB
	//options[2].optionString = "-Xmx1g"; // 1GB

	JavaVMInitArgs vm_args;
	memset(&vm_args, 0, sizeof(vm_args));

	vm_args.version = JNI_VERSION_1_6;
	vm_args.nOptions = 1;
	vm_args.options = options;
	vm_args.ignoreUnrecognized = JNI_TRUE;

	int result = JNI_CreateJavaVM(p_vm, p_env, &vm_args);
	LOGI("JNI_CreateJavaVM: %d", result);
	return result;
```
```
void RequestPermissions()
{
	JavaVM* vm = nullptr;
	JNIEnv* env = nullptr;
	int status = InitJVM(&vm, &env);
	if (status == 0)
	{
		LOGI("Initialization success (vm=%p, env=%p)", vm, env);
	}
	else
	{
		LOGI("Initialization failure (%i)", status);
		return;
	}
}
```

**Comment by enh** [comment]
you don't want to *create* a new VM... you want to use the existing one that you're already running in. that's passed to your `JNI_OnLoad` function.

**Comment by Gibson85** [comment]
Ok, that works now. But I have the next problem when calling a test function "IncDepth" of the "Permissions" java module:

```
                jclass clazz = env->FindClass("com/my/app/Permissions");

		jmethodID methodID = env->GetStaticMethodID(clazz, "IncDepth", "(I)I");
```

```
package com.my.app;

import android.Manifest;
import android.content.pm.PackageManager;
import android.support.v4.content.ContextCompat;
import android.util.Log;

public class Permissions
{
	public static int IncDepth(int depth) { return depth + 1; }
};
```

```
JNI DETECTED ERROR IN APPLICATION: JNI GetStaticMethodID called with pending exception java.lang.ClassNotFoundException: Didn't find class "com.my.app.Permissions" on path: DexPathList[[directory "."],nativeLibraryDirectories=[/system/lib, /vendor/lib, /system/lib, /vendor/lib]]
2019-07-02 20:03:13.726 32001-32054/com.my.app A/art: art/runtime/java_vm_ext.cc:470]   at java.lang.Class dalvik.system.BaseDexClassLoader.findClass(java.lang.String) (BaseDexClassLoader.java:56)
2019-07-02 20:03:13.726 32001-32054/com.my.app A/art: art/runtime/java_vm_ext.cc:470]   at java.lang.Class java.lang.ClassLoader.loadClass(java.lang.String, boolean) (ClassLoader.java:380)
2019-07-02 20:03:13.726 32001-32054/com.my.app A/art: art/runtime/java_vm_ext.cc:470]   at java.lang.Class java.lang.ClassLoader.loadClass(java.lang.String) (ClassLoader.java:312)
2019-07-02 20:03:13.726 32001-32054/com.my.app A/art: art/runtime/java_vm_ext.cc:470] 
2019-07-02 20:03:13.726 32001-32054/com.my.app A/art: art/runtime/java_vm_ext.cc:470]     in call to GetStaticMethodID
```

**Comment by enh** [comment]
that looks like you're using the wrong class loader. if you call `FindClass` in `JNI_OnLoad` you'll get the right class loader, the one that knows where to find your app's code. see https://developer.android.com/training/articles/perf-jni#java (in particular the last paragraph). my guess is that your two-line code snippet is running on some random native thread that you've attached with `AttachCurrentThread`, and so it doesn't have any association with the specific class loader that you need.

TL;DR: do the `FindClass` in your `JNI_OnLoad` too.

**Comment by Gibson85** [comment]
Right, I attached a random thread - my main thread.
I start from "ANativeActivity_onCreate" on to integrate Android into my framework.
Also not using native_app_glue.

I did all this for not needing this useless "JNI_OnLoad" function.
Then I think I should go to create my own JavaVM as I tried before.
Please could you provide a working example or tell me what options are needed
to tell the JavaVM about my java files.

```
	JavaVMOption options[4];
	memset(&options, 0, sizeof(options));

	options[0].optionString = "-Djava.class.path=.";
        ....
	options[3].optionString = "-verbose:jni";
```
As stated before "JNI_CreateJavaVM" always return "-1" unknown error.

That startup logic for android is a pain in the a...

**Comment by DanAlbert** [comment]
> That startup logic for android is a pain in the a...

Well, you are avoiding the tools that make it easy. If you don't want it to be a pain in the ass, why aren't you using `native_app_glue`?

If you can't use `native_app_glue`, it looks like the `ANativeActivity` struct that's passed to `ANativeActivity_onCreate` has the `JavaVM` you're looking for: https://developer.android.com/ndk/reference/struct/a-native-activity.html#struct_a_native_activity

> Please could you provide a working example or tell me what options are needed

https://github.com/googlesamples/android-ndk/tree/master/native-activity

**Comment by Gibson85** [comment]
Good point, thanks. My test code looks like this now:
```
static int InitializeJVM(JavaVM** vm, JNIEnv** env)
{
	ANativeActivity* activity = reinterpret_cast<ANativeActivity*>(GetANativeActivity());
	*vm = activity->vm;
	*env = activity->env;
	if ((*vm)->AttachCurrentThread(env, nullptr) != 0)
	{
		LOGI("Failed to attach current thread.");
	}
	jint ver = (*env)->GetVersion();
	LOGI("JVM load succeeded. Version %d.%d", ((ver >> 16) & 0x0f), (ver & 0x0f));
	return JNI_OK;
}
```
But when calling

```
		jclass clazz = env->FindClass("com/my/app/Permissions");
		jmethodID methodID = env->GetStaticMethodID(clazz, "IncDepth", "(I)I");
```
I am still getting the exception below.
So at the end I guess that code resultet in the same JavaVM as provided by the "JNI_GetCreatedJavaVMs" call.

```
2019-07-04 19:45:05.435 12990-13059/com.my.app A/art: art/runtime/java_vm_ext.cc:470] JNI DETECTED ERROR IN APPLICATION: JNI GetStaticMethodID called with pending exception java.lang.ClassNotFoundException: Didn't find class "com.my.app.Permissions" on path: DexPathList[[directory "."],nativeLibraryDirectories=[/system/lib, /vendor/lib, /system/lib, /vendor/lib]]
2019-07-04 19:45:05.435 12990-13059/com.my.app A/art: art/runtime/java_vm_ext.cc:470]   at java.lang.Class dalvik.system.BaseDexClassLoader.findClass(java.lang.String) (BaseDexClassLoader.java:56)
2019-07-04 19:45:05.435 12990-13059/com.my.app A/art: art/runtime/java_vm_ext.cc:470]   at java.lang.Class java.lang.ClassLoader.loadClass(java.lang.String, boolean) (ClassLoader.java:380)
2019-07-04 19:45:05.435 12990-13059/com.my.app A/art: art/runtime/java_vm_ext.cc:470]   at java.lang.Class java.lang.ClassLoader.loadClass(java.lang.String) (ClassLoader.java:312)
```

**Comment by Gibson85** [comment]
So to sum up the ANativeApplication struct provides a JavaVM/JNIEnv which can't access the java classes of the app. Is that right? I hope I am doing something wrong, because otherwise I have to call Android broken by design. For what is that JavaVM If it can't be used for nothing?

**Comment by enh** [comment]
> I did all this for not needing this useless "JNI_OnLoad" function.

i think you've now learned the hard way why `JNI_OnLoad` is very much not useless...

the last paragraph of https://developer.android.com/training/articles/perf-jni#native-libraries explains the class loader issues (but assumes that the reader understands class loaders quite well). TL;DR: avoid corner cases by using `JNI_OnLoad` for `FindClass` calls.
